docs: IOx Transactions & Locks (#3411)
influxdata/influxdb_iox: docs/transactions_locks.md

## Read and Write Physical Data Chunks
As described in [IOx Data Organization and LifeCycle](data_organization_lifecycle.md), there are four types of `Data Chunks` in IOx: `O-MUB`, `F-MUB`, `RUB`, and `OS` in which only `O-MUB` is mutable, the others are immutable. When a write is issued, it will go to its corresponding `O-MUB`, but when a read occurs, it may need to read all types of chunks that may contain needed data[^prune].

> [^prune]: IOx supports chunk pruning to eliminate chunks that do not contains needed data but it is beyond the scope of this document.

Reviewer: ```suggestion
[^prune]: IOx supports chunk pruning during query planning to eliminate chunks that do not contains needed data but it is beyond the scope of this document.
```

> * **Read a `F-MUB`, a `RUB` or a `OS`:** When an immutable data chunk is identified to be read, its reference will be returned without locking because it never gets modified. If IOx wants to remove this chunk, it will wait until no references to the chunk before doing so. Next section will describe how the catalog objects that lead IOx to this data chunk is locked.

Reviewer: ```suggestion
* **Read a `F-MUB`, a `RUB` or a `OS`:** When an immutable data chunk is identified to be read, its reference will be returned without locking because it never gets modified. If IOx wants to remove this chunk, it will wait until no references to the chunk exist before doing so. Next section will describe how the catalog objects that lead IOx to this data chunk is locked.
```

* A thread holds an **exclusive (or write) lock** on an object `T` means it is modifying `T` and no other threads are able to read or modify `T`. If `T` was snapshot by other threads before the write lock was acquired, those threads are still able to read the snapshots while `T` is being modified.

Table 1 explains that while `Thread 1` is holding a lock (`S` or `X`) on an object `T`, `Thread 2` can only acquired `S` lock on `T` if the one `Thread 1` is holding is `S`. The way we read the table:
* `S` lock is compatible with `S` lock (itself) but not compatible with `X` lock.
* `X` lock is not compatible with any locks.

```text

                     \    Thread 1 ->      |    S   |   X 
       Thread 2 |     \                    | (read) |(write)
                V      \                   |        |
    ───────────────────────────────────────|────────|────────
      Possible to acquire read lock (S)?   |  yes   |  no
    ───────────────────────────────────────|────────|────────
      Possible to acquire write lock (X)?  |  no    |  no

    
Table1 1: Common `Shared` and `Exclusive` Locks and Their Compatibility
```

> IOx supports one more lock in the middle, `SX`, as shown in Table 2 and implemented in [Freezable](https://github.com/influxdata/influxdb_iox/blob/fa47fb5582cb7527817a8c2834b82b5eb604ad46/internal_types/src/freezable.rs). `SX` is a lock that is first acquired for `read` but will be upgraded to `write` at some point. This allows other concurrent threads to read the `SX`-acquired object before it is upgraded to `write` but won't be able to acquire `SX` or `X` on that object. This means while `Thread 1` is holding `SX` on object `T`, other threads cannot modify T; and no matter how many threads are holding `S` lock on `T` while `Thread 1` is holding `SX` on it, `Thread 1` is still able to upgrade its lock to `X` for modifying `T`.

Reviewer: ❤️

In this case we say:
* `S` is compatible with itself and `SX`
* `SX` is only compatible with `S`
* `X` is only compatible with `S` if it is upgrading from an `SX`.

IOx also support locking on Rust futures which means a lock won't be actually acquired until its corresponding future happens. 

```text

                     \           Thread 1 ->           |     S               | SX (read-then-write) | X (write)
         Thread 2 |   \                                |  (read)             | (read-then-write)    |
                  V    \                               |                     |                      |
    ───────────────────────────────────────────────────|─────────────────────|──────────────────────|───────────
      Possible to acquire read lock (S)?               |     yes             |        yes           |   no
    ───────────────────────────────────────────────────|─────────────────────|──────────────────────|───────────

>       Possible to acquire read-then-write lock (SX) ?  |     yes?            |        no            |   no

Reviewer: I think it is possible to get a `SX` lock (Freeze handle) when there are other outstanding read locks so 👍  @tustvold  should confirm.

    ───────────────────────────────────────────────────|─────────────────────|──────────────────────|───────────
      Possible to acquire write lock (X)               | . yes if holding SX |        no            |   no         
                                                       | . no otherwise      |                      |
    
Table 2: IOx Locks and Their Compatibility
```

## Examples of Locks and Transactions
Let us go over a few common actions in IOx Data Lifecycle to explore when locks and transactions are used.

### Example 1: Query Data of a Table, T1
In order to query a table, we need through these steps:
1. Step 1: Identifying Data Chunks of the table and take their snapshots.
1. Step 2: Pruning unnecessary Chunks
1. Step 3: Building a query plan for those Chunks
1. Step 4: Executing the plan

> Since querying only involves reading activities, no transactions are needed. As pointed out above, locks no longer needed when snapshots or references of the Data Chunk are identified, only the first step acquires locks and the subject of discussion of this document.

Reviewer: ```suggestion
Since querying only involves reading activities, no transactions are needed. As pointed out above, locks are no longer needed once snapshots or references of the Data Chunk are identified, so only the first step acquires locks and the subject of discussion of this document.
```

**Step 1: Identify Data Chunks of the tables and take their snapshots** are implemented in [filtered_chunks](https://github.com/influxdata/influxdb_iox/blob/8a2410e161996603a4147e319c71e5bb38ca9cb7/server/src/db/catalog.rs#L288)[^tables] and as follows:

>   1. <span style="color:red">Acquire</span> `S` lock on all Catalog Tables.

Author: I am not sure why this color does not show up when I am on `display in rich diff`

  1. Identify Catalog Table of `T1` and, let say, its only appropriate Catalog Partition for this query, `P1`.
  1. <span style="color:red">Acquire</span> `S` lock on `P1`.
  1. Identify Catalog Chunks of `P1` which are, let say, `C1` and `C2`.
  1. <span style="color:red">Acquire</span> `S` lock on `C1` and `C2`.
  1. Take snapshots for `C1` and `C2` (snapshot can either be real snapshot for `O-MUB` or reference for `F-MUB`, `RUB` and `OS`)
  1. <span style="color:green">Release all</span> `S` locks on Chunks (`C1` and `C2`), Partition (`P1`), and Tables.
  1. Return the snapshots.

[^tables]: The filtered_chunks function can return chunks of many tables but we only need chunks of one table in this example.

### Example 2: Write to an O-MUB Chunk of Partition, P1, of Table, T1

> A write in IOx is implemented in [store_filtered_write](https://github.com/influxdata/influxdb_iox/blob/ccba68fe3ed3b41f06992cd1c11eefe720ed3ad4/server/src/db.rs#L1053) that can trigger to create a table or partitions as needed but in this example, we only focus on writing data to an available `O-MUB` chunk of Partition, `P1`, of Table, `T1`. Similar to reading, the lock acquisition and release are in the same order but on `X` lock instead.

Reviewer: ```suggestion
A write in IOx is implemented in [store_filtered_write](https://github.com/influxdata/influxdb_iox/blob/ccba68fe3ed3b41f06992cd1c11eefe720ed3ad4/server/src/db.rs#L1053) that can trigger the creation of a table or partitions as needed but in this example, we only focus on writing data to an available `O-MUB` chunk of Partition, `P1`, of Table, `T1`. Similar to reading, the lock acquisition and release are in the same order but on `X` lock instead.
```

1. Step a: <span style="color:red">Acquire</span> `S` lock on **cleanup lock** to ensure no cleanup jobs can remove any parquet files.
1. Step b: Write the compacted data to a parquet file, `PF`, in the object store.

> 1. Step c: <span style="color:red">Open</span>  a Catalog Transaction, `CT`

Reviewer 2: This acquires another lock, only a single catalog transaction can be being performed at once. 

As mentioned above, I would suggest splitting out details of preserved catalog transactions into their own document, and keep them separate from those on the in-memory catalog. Mixing the concepts together I think is a source of a lot of confusion
